File: cpp/base/src/vsdk/toolkit/common/logging/Logger.cpp
#include "vsdk/toolkit/java/lang/String.h"
#include "vsdk/toolkit/java/lang/System.h"
#include "vsdk/toolkit/common/logging/Logger.h"

/**
Prints an error message. Behaves much like printf. The first argument is the
name of the routine in which the error occurs (optional - can be nullptr)
*/
void
Logger::error(const char *routine, const char *text, ...) {
    va_list variableList;

    java::System::err.printf("Error: ");
    if ( routine ) {
        java::System::err.printf("%s(): ", routine);
    }

    va_start(variableList, text);
    const java::String message = java::String::formatCStringToJavaString(text, variableList);
    va_end(variableList);
    java::System::err.print(message.toCString());

    java::System::err.printf(".\n");
    java::System::err.flush();
}

/**
Fatal error: print message or and exit the program with the specified error code
First argument is a return code. We use negative return codes for
"internal" error messages
*/
[[noreturn]] void
Logger::fatal(int errcode, const char *routine, const char *text, ...) {
    va_list pvar;

    java::System::err.printf("logFatal error: ");
    if ( routine ) {
        java::System::err.printf("%s(): ", routine);
    }

    va_start(pvar, text);
    const java::String message = java::String::formatCStringToJavaString(text, pvar);
    va_end(pvar);
    java::System::err.print(message.toCString());

    java::System::err.printf(".\n");
    java::System::err.flush();

    java::System::exit(errcode);
}

/**
Same, but for warning messages
*/
void
Logger::warning(const char *routine, const char *text, ...) {
    va_list pvar;

    java::System::err.printf("Warning: ");
    if ( routine ) {
        java::System::err.printf("%s(): ", routine);
    }

    va_start(pvar, text);
    const java::String message = java::String::formatCStringToJavaString(text, pvar);
    va_end(pvar);
    java::System::err.print(message.toCString());

    java::System::err.printf(".\n");
    java::System::err.flush();
}

File: typescript/base/src/main/vsdk/toolkit/common/logging/Logger.ts
const util = require("node:util");

export class Logger {
  private constructor() {
  }

  public static error(routine: string | null, text: string | null, ...args: unknown[]): void {
    process.stderr.write("Error: ");
    if (routine !== null) {
      process.stderr.write(`${routine}(): `);
    }
    process.stderr.write(`${Logger.formatMessage(text, ...args)}.\n`);
  }

  public static warning(routine: string | null, text: string | null, ...args: unknown[]): void {
    process.stderr.write("Warning: ");
    if (routine !== null) {
      process.stderr.write(`${routine}(): `);
    }
    process.stderr.write(`${Logger.formatMessage(text, ...args)}.\n`);
  }

  public static fatal(errcode: number, routine: string | null, text: string | null, ...args: unknown[]): void {
    process.stderr.write("logFatal error: ");
    if (routine !== null) {
      process.stderr.write(`${routine}(): `);
    }
    process.stderr.write(`${Logger.formatMessage(text, ...args)}.\n`);
    process.exit(errcode);
  }

  private static formatMessage(text: string | null, ...args: unknown[]): string {
    if (text === null) {
      return "";
    }
    try {
      return util.format(text, ...args);
    }
    catch (_e) {
      return text;
    }
  }
}
